demos/shadow/shadow_demo_main/shadow_demo_main.c

```diff
@@ -0,0 +1,644 @@
+        JSONStatus_t result = JSONSuccess;
+
+        result = JSON_Search( ( char * ) pPublishInfo->pPayload,
+                              pPublishInfo->payloadLength,
+                              "version",
+                              sizeof("version") - 1,
+                              '.',
+                              & outValue,
+                              ( size_t * ) & outValueLength );
+        if ( result == JSONSuccess )
+        {
+            LogInfo( ( "version: %.*s\n\n",
+                        outValueLength,
+                        outValue ) );
+
+            /* Convert the code to an unsigned integer value. */
+            version = ( uint32_t ) strtoul( outValue, NULL, 10 );
+
+            LogInfo( ( "version:%d, currentVersion:%d \r\n", version,  currentVersion) );
+
+            if ( version > currentVersion )
+            {
+                /* Set to received version as the current version. */
+                currentVersion = version;
+
+                /* Get powerOn state from json documents. */
+                result = JSON_Search( ( char * ) pPublishInfo->pPayload,
+                                      pPublishInfo->payloadLength,
+                                      "state.powerOn",
+                                      sizeof("state.powerOn") - 1,
+                                      '.',
+                                      & outValue,
+                                      ( size_t * ) & outValueLength );
+
+                if ( result == JSONSuccess )
+                {
+                    /* Convert the code to an unsigned integer value. */
+                    newState = ( uint32_t ) strtoul( outValue, NULL, 10 );
+
+                    LogInfo( ( "The new power on state newState:%d, currentPowerOnState:%d \r\n",
+                               newState,  currentPowerOnState) );
+
+                    if ( newState != currentPowerOnState )
+                    {
+                        currentPowerOnState = newState;
+
+                        /* State change will be handled in main(), where we will publish a "reported"
+                        * state to the device shadow. We do not do it here because we are inside of
+                        * a callback from the MQTT library, so that we don't re-enter
+                        * the MQTT library (even though the library might support re-entrance.) */
+                        stateChanged = true;
+                    }
+                }
+                else
+                {
+                    LogError( ( "No powerOn in json document!!\n\n" ) );
+                }
+            }
+            else
+            {
+                LogError( ( "The received version is smaller than current one!!\n\n" ) );
+            }
+        }
+        else
+        {
+            LogError( ( "No version in json document!!\n\n" ) );
+        }
+    }
+    else
+    {
+        LogError( ( "The json document is invalid!!\n\n" ) );
+    }
+}
```

Comment:
I begin to see why the complexity score you listed is so high.  You might consider breaking this into three functions.

Reply:
Since it's a demo code, I thought it's would be reasonable to make the user much more aware how to handle these messages than splitting into sub functions to decrease the complexity.

Reply:
will address the deep nesting.

Reply:
[This commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1104/commits/1c5938bf39c086c340dcc4e83dedc5c2482cf7a7) reduce the deep nesting.